src/main/scala/is/hail/stats/RegressionUtils.scala

```diff
@@ -70,16 +70,77 @@ object RegressionUtils {
     (y, cov, completeSamples)
   }
 
-  def buildGtColumn(gts: Iterable[Option[Int]]): Option[DenseMatrix[Double]] = {
-    val (nCalled, gtSum, allHet) = gts.flatten.foldLeft((0, 0, true))((acc, gt) => (acc._1 + 1, acc._2 + gt, acc._3 && (gt == 1) ))
+  def mutateLastColumn(X: DenseMatrix[Double], gts: IntIterator, mask: Array[Boolean]): Boolean = {
+    require(X.offset==0 && X.majorStride==X.rows && !X.isTranspose)
 
-    // allHomRef || allHet || allHomVar || allNoCall
-    if (gtSum == 0 || allHet || gtSum == 2 * nCalled || nCalled == 0 )
+    val n = X.rows
+    val k = X.cols - 1
+    var missingIndices = new mutable.ArrayBuilder.ofInt()
+    var i = 0
+    var j = k * n
+    var gtSum = 0
+    while (i < mask.length) {
+      val gt = gts.nextInt()
+      if (mask(i) && gt != -1) {
+        gtSum += gt
+        X.data(j) = gt.toDouble
+        j += 1
```

Comment:
Ah, you're right.

The code now seem strictly worse than what you just wrote since it checks mask(i) twice.

if (a && b) ...

is identical to

if (a) if (b) ...